### ISSUE TITLE
#bank sections and index.bas not always building with RUNinCSpect v1.11.1

### ISSUE THREAD

**Issue by NealeTools** [remy/vscode-nextbasic#47]
Hi Remy - sorry...another bug (for me). Sometimes the files with #bank directives don't seem to build.
For example: Initially, RUN in CSpect just kept loading a pre-exisitng old index.bas.
Eventually, after tying to hash out the #bank part, and then readd it,  I got it working with a simple demo with one added #BANK directive.
Just now I added a second #bank BANK21, but now it just keeps loading up the original .bas
Any ideas? I've saved the file, but it seems to make no difference.

1st Example:
<img width="993" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/b8a35af3-8ad4-455b-adb6-46b1532ea787">

Modified:
<img width="593" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/34a36b25-24ba-410b-91c5-b24ce1791d93">

Now if I hash out those #bank directives, it finally loads up the file (that must have been made from a prior run?)
<img width="590" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/01e64683-eda3-4641-9c71-04b9a966e340">

I then tried adding a third #bank and this happened:
<img width="604" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/926c82b6-8211-4061-adf6-230929d66860">

So I hashed out the final #bank: 
<img width="593" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/f301d604-07a2-402a-ae05-07abdd13a294">
This ought to crash on line 37 because BANK 22 isn't defined, but it just doesn't go there. It isn't loading the BASIC with line 37 in it: 
<img width="299" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/2e5a785f-bdf1-4d89-b8d5-705237dead01">

Now, it finally works again (after unhashing):
<img width="594" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/855d6272-d362-4b22-bb8a-ba2cf254f2d5">

Any idea what I am doing wrong here? Is it something to do with file saving? (Is that necessary? Maybe it is a bit slow or something, or not seeing the new file?)

**Comment by NealeTools** [comment]
I also had  a look at the current (last) index.bas file, assuming is still how this works. It has no line numbers in it and looks incomplete, is that normal?
<img width="587" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/d0625153-088c-4c29-8b5a-84ceb4a34b50">

**Comment by NealeTools** [comment]
When I get errors, I think some of the files are not being created correctly - perhaps not overwriting.
As an example, iside the image I see:
<img width="765" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/49cdd2d5-1100-4e6a-a66a-c600bbdf3616">

**Comment by remy** [comment]
Not gone through these yet, but  do not apologise - you're literally helping by testing! These issues are _really_ useful.

**Comment by NealeTools** [comment]
Both BANK22 are "identical" as viewed in macOS:
<img width="551" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/03395ae8-68b3-42c7-94df-4d7f5ee50455">

I then tried to open them in textedit. Remarkably, when I tried to do this one of the BANK22 files vanished!
<img width="1200" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/572f980f-52f5-4156-9024-09c02aa33b02">

**Comment by NealeTools** [comment]
ALl is weird...I now added some PAUSE 0 to test. Retried building. It worked this time!
But this isn't the new code with PAUSE 0 in there! I assume it is the prior in index.bas that is loading up!
<img width="549" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/1552a01d-af41-40e9-aeea-75fe4101cb76">
<img width="321" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/5b257f33-1c18-40a8-af42-541b621aedf5">
I'm at a loss...

Even more testing: Now all the BANKs are behaving as if they have the same code in them!...
<img width="555" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/e2a4c598-928e-4c91-aab6-b5d7c5df6142">

**Comment by NealeTools** [comment]
OK. I did a quick test with a fresh .img—just to be sure it wasn't a problem with my extremely full devel folder. I emptied and then tried again. This time it failed again, and inspecing the .img shows:
<img width="927" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/0c6a78dc-c4fc-4220-8557-dfe4410db0ac">

Interestingly, trying to load TirstTest.bas gives Wrong File Type 0:2 error:
<img width="312" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/656fd027-862b-4623-baec-b82d39b20e6f">
Yet, when dragged into your web tool, it looks good...

<img width="625" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/328aa972-c252-43c8-b92e-29197ba068ca">

**Comment by remy** [comment]
The web tool is a couple of fixes behind, but also I don't think it knows how to handle bank splitting (partly because I was sure how I was going to do it yet!).

So that might actually be a parsing bug (in the txt2bas).

**Comment by NealeTools** [comment]
I wasn't trying to use the web tool to split the file. The index.bas is already tokenised. Given that it was apparently not loading in NextBASIC, I just wanted to see if your (older) webtool could convert it back to .txt and recognise the tokens (it can). I'm not sure if this provides more insight or not...

**Comment by remy** [comment]
Found it - or, at least I've made a fix that can't replicate, and the bug I fixed, I'm fairly sure is responsible for the confusion.

If you can read JavaScript, maybe you can spot it: https://github.com/remy/vscode-nextbasic/blob/master/lib/auto-run.task.js#L179-L197

Otherwise, basically, the main program wasn't being synced properly, yet the banks, kinda were.

There's a lot in your issue, so I'm hoping re-testing isn't too bad, and hopefully the fix does the job.

The latest vscode extension version is 1.11.2 which includes fixes to all three issues raised.

**Comment by NealeTools** [comment]
I can't read javascript very well... ;-)
ChatGPT says:
In the newer version of the code, there's an adjustment related to the file2bas function call, specifically the addition of the bankOutputDir argument. Let's take a closer look at this change:

In the newer version:

javascript
Copy code
writeFileSync(
  filename,
  file2bas(text, {
    validate: false,
    defines: true,
    bankOutputDir: tmpdir(), // <- New argument added here
  }),
  'utf8'
);
In the older version:

javascript
Copy code
writeFileSync(
  filename,
  file2bas(text, {
    validate: false,
    defines: true,
  }),
  'utf8'
);
In the newer version, the bankOutputDir argument is added to the options object passed to the file2bas function. This argument specifies the directory where the output files (banks) generated by the file2bas function should be placed. In this case, it's set to tmpdir(), which is a function from the Node.js os module that returns the operating system's default directory for temporary files.

By adding the bankOutputDir argument, the code instructs the file2bas function to output any generated files (banks) to the specified directory. This can be useful for organizing files and managing the output of the file2bas function, especially when dealing with multiple files or complex projects.

This change allows for more flexibility and control over where the output files are stored, potentially improving the organization and management of files within the application.

:-)

**Comment by NealeTools** [comment]
Great progress! Unfortunately, I still get an error (not a crash). It seems that contents of BANK 22 and BANK 20 in my test are identical:
<img width="567" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/ac74be64-5d1b-45b9-be37-7c22ffd95cc8">

**Comment by NealeTools** [comment]
Inspecting the .img shows BANK22 duplicated:
<img width="756" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/b72816ca-28fd-48ba-92ac-1f6762be264b">
So I deleted them and tried again.
First try failed with file not found.
I tried again. This time it runs fine:
<img width="564" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/022c79c9-a37a-47ef-8442-79f321940769">

Can it be that there is some delay in the files being written into the .img?

**Comment by NealeTools** [comment]
Hmmm...I'd say that 9/10 times it crashes due to a malformed file that it can't load:
<img width="566" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/46d04c84-17b8-4da2-bdcb-d01a336fac0a">
:-(

**Comment by NealeTools** [comment]
Sometimes even the index.bas won't load and it crashes to the autoexec.bas:
<img width="561" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/538a5fb0-5e5b-4654-91c1-de1123dcf06c">

**Comment by remy** [comment]
So, replies:

- ChatGPT was wrong, and didn't spot the bug 😂 (doesn't suprise me though)
- re: delay writing to the img: no, the `put` calls are synchronous so cspect isn't launched until it's all across
- I'm wondering if there's actually two things going on...some syntax bug plus something with the banks, but nothings jumping out at me...

Can I ask the following:

- empty out your `/devel` directory from the img (or better yet, remove it entirely).
- restart your test and stop immediately on the first failure.
- mount the .img file with Finder and copy out the contents of the new `/devel` directory, zip the files up and send them across.

I can then inspect the raw byte data and hopefully also use it to replicate on my side.

**Comment by remy** [comment]
Wait, just saw your duplicated files, I've no idea how that can happen...it's literally a file system that should prevent that from happening... (that's confused me a lot... )

**Comment by NealeTools** [comment]
OK - I deleted devel. I just retried the test. It broke after loading up the index.bas:
<img width="593" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/5bcceb47-ae86-4691-98ad-6f2eed306762">

I remounted .img and inspected. devel is back...but it is empty...how is that even possible, if CSpect loaded the index.bas?
Is it loading from a cache?
<img width="870" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/2c7963e0-9c07-4cc9-8351-a862019e5914">

**Comment by remy** [comment]
Is the image mounted whilst you're running cspect? It can't be, it causes weird file system effects. If it wasn't, I might be inclined to get a new image file and start afresh (I did this last week myself)

**Comment by NealeTools** [comment]
It is sometimes mounted, sometimes not. I've never had any problem with that. It doesn't update in macOS once mounted. So to see changes made by CSpect, I have to trash and remount.

I just retired the test. It failed. So I went to browser to inspect...and this happened...
<img width="580" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/2db5a801-379a-44be-a250-52bd273a409e">

**Comment by NealeTools** [comment]
Here is the most recent attempt. One of the BANK files never got created, and one is zero bytes:
[devel.zip](https://github.com/remy/vscode-nextbasic/files/14531623/devel.zip)

**Comment by remy** [comment]
Cheers. I'll take a look at the zip in the morning. 

I've had issues the whole time with mounting the .img file, often the system would eventually become read only and files wouldn't write properly, so I've trashed the .IMG a good number of times over the years. 

I'll let you know if I find anything useful in there (I'm starting to suspect logs in the extension might be the way to debug though...)

**Comment by NealeTools** [comment]
OK - I just downloaded a new distro image, trashed the entire contents and replaced with the GitLab 2.08 verion. Ran my script and...
<img width="575" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/244b4ddb-bafd-4ee7-9438-92a192a9e659">

With new devel looking like this again:
<img width="756" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/cecc7604-4ef0-433f-bf05-da83fb168a61">

BANK21 fails to build...

**Comment by remy** [comment]
<img width="1162" alt="SCR-20240308-iocw" src="https://github.com/remy/vscode-nextbasic/assets/13700/90dd97bc-acfe-414c-9e35-3f68482f34cc">

Right. I'm there. I can replicate. That's the hard part down. Now to find out, what is actually happening!

**Comment by remy** [comment]
I'll add the auto incrementing lines is also hosed with split banks, as is the line number order validation - but I'll keep that as a separate issue.

**The distilled issue here is** (for future me)

- split banks are copied across to the cspect image and resulting in zero bytes, or not at all.

I've checked $TMPDIR and the banks are there with correct sizes, but they're definitely not syncing properly.

**Comment by remy** [comment]
Sod, also just spotted this where:

> no, the put calls are synchronous so cspect isn't launched until it's all across

However, I've just spotted the individual banks are actually using an async (whereas nothing else is). That's a bug, it could also explain what's going on. Just checking.

**Comment by remy** [comment]
![SCR-20240308-itcf-min](https://github.com/remy/vscode-nextbasic/assets/13700/1f31516a-6229-4c8e-bd83-d5f5b546ab9f)

Think I've got it.

**Comment by remy** [comment]
Right, 1.11.3 - give that a test, I think that's it!

(I'll file the syntax highlighting for bank split line numbers separately)

**Comment by NealeTools** [comment]
v1.11.2...A new crash I 'v never seen before lol!
<img width="636" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/670803e7-ee2b-4ee7-a5cb-790611fdf515">

v1.11.3: Oh yeah! :-)
<img width="547" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/c2fdfdab-8097-4439-9767-8f7719e983d5">

**Comment by NealeTools** [comment]
Nice! And later references to the same FILENAME over-write, rather than crash.
I think this is OK...unless you particularly want to try to catch this?
<img width="562" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/cdc5cafe-a56d-4967-b2f4-ec8cd1ef506f">

**Comment by remy** [comment]
I had that same bizarre stripy error too - I'm pretty sure it was loading random data from an "empty"/junk bank. First time I'd seen it before too!

**Comment by NealeTools** [comment]
This really is fantastic - thanks so much for taking the time developing this tool (not to scare you off, but 2.09 is very soon to launch...).
Now...about the export feature: As I think I said before, I previously modified the .js file so that it additionally copied the resulting index.bas to a designated folder that I am syncing to the Next HW over wifi .sync. I quite like that process because it means I have a single workflow that I can use for both in parallel should I choose. I guess that to do this now, will be a littel more complicated, and I think I might need help. Could an additional export path be defined in the VSCode settings section of the Extension?

Easily modifiable that way, and I'm sure you would also know how to correctly write the javascript so that the same files were copied to both the .img and to this additional path.

**Comment by remy** [comment]
I'm not sure what your workflow is, but I've been doing this myself for the games I wrote using the export functionality.

I also sneaked in key binding (cmd+shift+e) will bring up the dialog box, though it default to the working directory, you can point it to any directory and it'll export all the files to that location.

**Comment by NealeTools** [comment]
Hi Remy - I've just tried this out, but I suspect there may be a small bug remaining that I didn't notice before. The export function created what appear to the the correct files in the working directory (GeneralCode) - top window.
FirstTest.bas is actually the master .txt file so ignore that.
Untitled.bas should the the master/main .bas prog (because I didn't give it a name), and the BANKs are the split off files.
Note, however, that the BANK files are recognised as .txt files by macOS (this is useful...but unusual).

In the lower window, I did a RunInCSpect version—and what you can see is a view in the remounted devel folder.
As you can see the BANK files are now listed as Unix executable files. This is what I normally see. I'm not sure if this difference matters, or why there would be a difference if the files are identical.
Secondly, for unknown reasons, "Untitled.bas" is also present in the devel folder. Where did it come from?
I'd cleared out the devel folder to be empty, and remounted to confirm it was empty before running the RunInCSpect macro.
<img width="1131" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/4ceeac24-3c40-49fc-9739-c26c9d166a60">

**Comment by remy** [comment]
So, I believe all this is actually fine.

1. Your mac picking up that the BANK1 is a text file is entirely based on your finder settings. I mean, it's wrong to think it's a text file, but the mac isn't actually looking at the file's contents, and only the file association.
2. Everything in the working directory is synchronised across to the the /devel directory when you test with cspect - so if you exported anything into the working directory (including sub-directories), it'll go across to cspect
3. For my own personal dev, I keep my plain text (source) files clearly labelled as text (where you're naming them as .bas - which is actually a binary format). For example, I would have FirstTest.bas.txt (so I know that this is source and not something I can drop into a spectrum, but also can open with any text editor). I also tend to organise the directory so that there's two sub-directores: `src` and `dist` - `src` containing the source text files, and `dist` contains the generated binary .bas (and banks). But, here's really what you prefer.

Finally as for the byte differences between Untitled.bas and your index.bas - if you zip them up I can tell you what they are. Though you can drop the index.bas into VS Code, it'll say "The file is not displayed in the text editor because it is either binary or uses an unsupported text encoding." - but open it anyway, then in VS Code's command palette, select "NextBASIC: import binary .bas to text" and it'll show you the bas2txt source.

**Comment by NealeTools** [comment]
Thanks Remy!
1. Ah - yes, but any idea why it detects some of the BANK files as unix executables (the ones generated in the .img) and some as if they are text (the ones generated by the export)?
2. Of course! 
3. Yes - the reason I named them .bas was a workflow fudge so that I could specifically associate by BAS text files with VSCode (for auto-opening), but not for all .txt files (which I prefer to open in TextEdit). It's not great...and probably I should make a bespoke suffix. I may see if .bas.txt be associated globally with VSCode in macOS.

As for index.bas vs untitled.bas: Quite a few byte differences...
<img width="1126" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/fe1cd317-a37f-4a66-8f21-35db0a6c47e1">

And after bas2txt:
<img width="1064" alt="image" src="https://github.com/remy/vscode-nextbasic/assets/19664927/8a91d1ee-c6b6-4783-aeba-5d3907e006fc">

Aha! It looks like autoline numebring is enabled in the Export ("Untitled") option! This explains the byte differences. :-)

**Comment by remy** [comment]
Of course. Yes, index.bas has autostart because it needs to launch when developing, but if you want that in the file, you actually need to define it (same way as you would with .txt2bas)

**Comment by NealeTools** [comment]
Ah - OK, so I just add: "#autostart 1" at the beginning of the code.
However, how would I turn off renumbering of the exported version?

**Comment by remy** [comment]
Renumbering is opt-in in your code. Are you seeing it happening and changing your source numbers? If that's the case, can we raise a new issue with example code?

**Comment by remy** [comment]
Ah, I'm only seeing the screenshots now (I get the messages in email, but not the images).

Let's open a new issue with the autoline thing. 

(just a note, if you .bas file contains autostart, it'll add those lines it in when importing from bas to text)

**Comment by NealeTools** [comment]
I raised a new issue #51
